test/prober/verify.go

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2020 The Knative Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prober
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/eventing/test/lib"
+	"math/rand"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+func (p *prober) Verify() ([]error, int) {
+	hostname := p.figureOutClusterHostname()
+	report := p.fetchReceiverReport(hostname)
+	p.log.Infof("Fetched receiver report. Events propagated: %v. "+
+		"State: %v", report.Events, report.State)
+	if report.State == "active" {
+		panic(errors.New("report fetched to early, receiver is in active state"))
+	}
+	errs := make([]error, 0)
+	for _, t := range report.Thrown {
+		errs = append(errs, errors.New(t))
+	}
+	return errs, report.Events
+}
+
+func (p *prober) Finish() {
+	p.removeSender()
+}
+
+type nodeAddresses []*corev1.NodeAddress
```

Comment:
I must be missing some background here, but why do we need the functions like `figureOutClusterHostname`, `prioritizeAddresses` here?

Reply:
@cardil Is it because the receiver deploys only Pod and Service and is not accessible from outside the cluster so you pick a random node and access the Service from within the cluster? In that case it would be better to just expose the Service through Route and get its address directly. No need to use Nodes.

Reply:
I'm making an HTTP GET request to receiver service, that has node port.

To do that I need a public address of one of the worker nodes. Those functions choose that address be looking at kubernetes nodes and choosing their addresses by priority. In htta way we have a public IP of one of the worker nodes.

Reply:
OK. Is there a specific reason for using type NodePort? Otherwise, see my comment https://github.com/knative/eventing/pull/2388#discussion_r368494179 for simplifying it.

Reply:
Let's discuss in one place.

@mgencur written:

> This could be ClusterIP in which case you need to expose it via Route or it could also be LoadBalancer and then you can get its hostname directly similar to what we do for getting istio-ingressgateway's hostname. No need to work with Nodes (see my other comment)

I used Minikube for most of the development. Route is OpenShift specific. I need to make HTTP request from outside the cluster. Adding LoadBalancer or using Ingress was in my mind more complicated, the using NodePort.

Reply:
OK. On Minikube it should be pretty simple to use LoadBalancer now as well, see: https://kubernetes.io/docs/tutorials/hello-minikube/#create-a-service  . The address of the service can be printed via `minikube service <service_name> --url` but then it needs to be passed to the program.

Reply:
I can use `LoadBalancer`, but that would not be simpler than using just `NodePort`. I've tested that on GCP, and Minikube. @mgencur looked it at OCP.

* On minikube I must use `NodePort`. I could use `minikube ip`, or the code that I wrote to search for external IP of node. 
* On GCP I can use `LoadBalancer`, but I must wait a couple of minutes, until external IP is assigned. Then I can use `.status.loadBalancer.ingress[0].ip` of a service
* On OCP (at AWS) I can also use `LoadBalancer`. Then I can use `.status.loadBalancer.ingress[0].hostname` of a service

I think using `NodePort` is more general solution here. I think that I can move the code, that lookup a node external IP to somewhere more general, like http://github.com/knative/test-infra as @mgencur pointed out.

Reply:
I've moved the code to https://github.com/knative/pkg/pull/1001